Break inside defer error

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.2.2 `47bf644`, timestamp: 2021-07-09 03:27:16 +0300
**OS:** Linux, Ubuntu 20.04.2 LTS

**What did you do?**

```
fn main() {
  for true {
    defer{
      break
    }
  }
}
```

**What did you expect to see?**

An error message about invalid statement inside defer's topmost block as a non-internal error
 
**What did you see instead?**

```
==================
/tmp/v/main.18109983032030113796.tmp.c:10183: error: cannot break
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -4468,9 +4468,6 @@ fn (mut c Checker) stmt(node ast.Stmt) {
 			c.inside_const = false
 		}
 		ast.DeferStmt {
-			if c.inside_defer {
-				c.error('defers are not allowed in defer statements', node.pos)
-			}
 			if node.idx_in_fn < 0 {
 				node.idx_in_fn = c.table.cur_fn.defer_stmts.len
 				c.table.cur_fn.defer_stmts << unsafe { &node }

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -74,6 +74,7 @@ pub mut:
 	inside_const   bool
 	inside_anon_fn bool
 	inside_ref_lit bool
+	inside_defer   bool
 	inside_fn_arg  bool // `a`, `b` in `a.f(b)`
 	inside_ct_attr bool // true inside [if expr]
 	skip_flags     bool // should `#flag` and `#include` be skipped
@@ -4467,6 +4468,9 @@ fn (mut c Checker) stmt(node ast.Stmt) {
 			c.inside_const = false
 		}
 		ast.DeferStmt {
+			if c.inside_defer {
+				c.error('defers are not allowed in defer statements', node.pos)
+			}
 			if node.idx_in_fn < 0 {
 				node.idx_in_fn = c.table.cur_fn.defer_stmts.len
 				c.table.cur_fn.defer_stmts << unsafe { &node }
@@ -4497,7 +4501,9 @@ fn (mut c Checker) stmt(node ast.Stmt) {
 					node.defer_vars[i] = id
 				}
 			}
+			c.inside_defer = true
 			c.stmts(node.stmts)
+			c.inside_defer = false
 		}
 		ast.EnumDecl {
 			c.enum_decl(node)
@@ -4542,6 +4548,9 @@ fn (mut c Checker) stmt(node ast.Stmt) {
 		}
 		ast.GotoLabel {}
 		ast.GotoStmt {
+			if c.inside_defer {
+				c.error('goto is not allowed in defer statements', node.pos)
+			}
 			if !c.inside_unsafe {
 				c.warn('`goto` requires `unsafe` (consider using labelled break/continue)',
 					node.pos)
@@ -4604,6 +4613,9 @@ fn (mut c Checker) block(node ast.Block) {
 }
 
 fn (mut c Checker) branch_stmt(node ast.BranchStmt) {
+	if c.inside_defer {
+		c.error('`$node.kind.str()` is not allowed in defer statements', node.pos)
+	}
 	if c.in_for_count == 0 {
 		c.error('$node.kind.str() statement not within a loop', node.pos)
 	}

--- a/vlib/v/checker/tests/defer_in_for.out
+++ b/vlib/v/checker/tests/defer_in_for.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/defer_in_for.vv:4:13: error: `break` is not allowed in defer statements
+    2 |     for true {
+    3 |         defer {
+    4 |             break
+      |             ~~~~~
+    5 |         }
+    6 |     }

--- a/vlib/v/checker/tests/defer_in_for.vv
+++ b/vlib/v/checker/tests/defer_in_for.vv
@@ -1,0 +1,7 @@
+fn main() {
+    for true {
+        defer {
+            break
+        }
+    }
+}

